File: VladislavShalnev/Term 2/Block 3/Task 6 (IoC)/TomorrowIoApi/TomorrowIo.cs
using System.Globalization;
using System.Runtime.CompilerServices;
using Newtonsoft.Json.Linq;
using Requests;
using TomorrowIoApi.Exceptions;
using TomorrowIoApi.Models.Timelines;
using WeatherApi;

[assembly: InternalsVisibleToAttribute("TomorrowIoApi.UnitTests")]
namespace TomorrowIoApi;

public class TomorrowIo : AWeatherApi
{
	protected override string BaseUrl => "https://api.tomorrow.io/v4";
	protected override string TokenName => "apikey";
	public override string Name => "TomorrowIo";

	private static readonly string[] DefaultFields = {
		"temperature",
		"cloudCover",
		"humidity",
		"precipitationIntensity",
		"windSpeed",
		"windDirection"
	};
	
	public TomorrowIo(string token) : base(token) {}

	// For tests only
	internal TomorrowIo(string token, IRequestsClient client) : base(token, client) {}

	public async Task<List<TimelineModel>> GetTimelinesAsync(
		(double lat, double lon) location,
		string[] timesteps,
		string[]? fields = null,
		string units = "metric" 
	)
	{
		string lat = location.lat.ToString(CultureInfo.InvariantCulture);
		string lon = location.lon.ToString(CultureInfo.InvariantCulture);
		
		fields ??= DefaultFields;
		
		var options = new Dictionary<string, string>
		{
			["location"] = $"{lat},{lon}",
			["fields"] = string.Join(",", fields),
			["timesteps"] = string.Join(",", timesteps),
			["units"] = units
		};

		JToken timelinesJson = (await CallMethodAsync("timelines", options))["data"]!["timelines"]!;
		List<TimelineModel>? timelines = timelinesJson.ToObject<List<TimelineModel>>();
		return timelines ?? new List<TimelineModel>();
	}

	public override Task<Weather> GetCurrentAsync((double lat, double lon) location) =>
		GetCurrentAsync(location);
	
	public async Task<Weather> GetCurrentAsync(
		(double lat, double lon) location,
		string[]? fields = null,
		string units = "metric"
	)
	{
		IntervalModel? interval = (await GetTimelinesAsync(
			location,
			new []{ "current" },
			fields,
			units
		))[0].Intervals?[0];
		
		return interval?.ToWeather() ?? new();
	}
	
	protected override void HandleErrors(JObject responseJson)
	{
		if (responseJson["code"] is not null)
			throw new ApiException(
				(int)responseJson["code"]!,
				(string)responseJson["type"]!,
				(string)responseJson["message"]!
			);
	}
}

File: ShangareevRustam/Second term/Block 4/1. Weather graphics app/lib/tomorrowIo/src/main/kotlin/tomorrowIo/TomorrowIo.kt
package tomorrowIo

import org.apache.http.impl.client.CloseableHttpClient
import weatherUtilities.Location
import weatherUtilities.UnassembledWeatherDataField
import weatherUtilities.WeatherCharacteristics
import weatherWeb.WeatherWeb

class TomorrowIo(
	override val appId: String,
	override val httpClient: CloseableHttpClient,
) : WeatherWeb() {

	override val url = "tomorrow.io"

	override val dataJSONInformation: Map<WeatherCharacteristics, UnassembledWeatherDataField> =
		mapOf(WeatherCharacteristics("Precipitation type") to UnassembledWeatherDataField(units = "",
			path = listOf("data", "timelines", "intervals", "values", "weatherCode")),
			WeatherCharacteristics("Precipitation") to UnassembledWeatherDataField(units = " mm/h",
				path = listOf("data", "timelines", "intervals", "values", "precipitationIntensity")),
			WeatherCharacteristics("Temperature") to UnassembledWeatherDataField(units = " °C",
				path = listOf("data", "timelines", "intervals", "values", "temperature")),
			WeatherCharacteristics("Cloud cover") to UnassembledWeatherDataField(units = " %",
				path = listOf("data", "timelines", "intervals", "values", "cloudCover")),
			WeatherCharacteristics("Humidity") to UnassembledWeatherDataField(units = " %",
				path = listOf("data", "timelines", "intervals", "values", "humidity")),
			WeatherCharacteristics("Wind direction") to UnassembledWeatherDataField(units = " °",
				path = listOf("data", "timelines", "intervals", "values", "windDirection")),
			WeatherCharacteristics("Wind speed") to UnassembledWeatherDataField(units = " m/s",
				path = listOf("data", "timelines", "intervals", "values", "windSpeed")))

	override fun getRequestUrl(location: Location) =
		"https://api." + url + "/v4/timelines?&timesteps=current&" +
				"fields=temperature,cloudCover,humidity,weatherCode," +
				"precipitationIntensity,windDirection,windSpeed&" +
				"location=${location.lat},${location.lon}" + "&units=metric" +
				"&apikey=$appId"
}